Multiple databases v2

calebhearth opened the pull request:
Add quite a bit more testing to the multiple databases feature, especially
around configuraiton, definition, and statement routing. Introduce a new
FakeAdapter class to let us test against a non-Postgres adapter without having a
dependency on any of the various actual adapters out there (we love you, but we
don't want a cyclic dependency!)

- Rework the new README changes to somewhat closer match the cheeky style of the
original.
- Fix array handling for migrations_paths. Rails lets us define a
`:migrations_paths` key that can be an array or a string in `database.yml`,
and we previously ignored the possibility of that being an array. New
functionality matches what Rails does.

Introduce configurable view paths! I'm hopeful that a lot of folks will be happy
with this. You can configure this with `views_paths` in `database.yml` and it
should default similarly to how migrations do to `db/<database_name>_views/` for
non-default databases.

Fix: https://github.com/scenic-views/scenic/issues/291
Fix: https://github.com/scenic-views/scenic/issues/380
Fix: https://github.com/scenic-views/scenic/issues/261
Resolves: https://github.com/scenic-views/scenic/issues/402
Closes: https://github.com/scenic-views/scenic/pull/451

dnamsons committed:
Feat: Multiple database support

calebhearth committed:
Multiple databases v2

Add quite a bit more testing to the multiple databases feature, especially
around configuraiton, definition, and statement routing. Introduce a new
FakeAdapter class to let us test against a non-Postgres adapter without having a
dependency on any of the various actual adapters out there (we love you, but we
don't want a cyclic dependency!)

- Rework the new README changes to somewhat closer match the cheeky style of the
  original.
- Fix array handling for migrations_paths. Rails lets us define a
  `:migrations_paths` key that can be an array or a string in `database.yml`,
  and we previously ignored the possibility of that being an array. New
  functionality matches what Rails does.

Introduce configurable view paths! I'm hopeful that a lot of folks will be happy
with this. You can configure this with `views_paths` in `database.yml` and it
should default similarly to how migrations do to `db/<database_name>_views/` for
non-default databases.

Fix: https://github.com/scenic-views/scenic/issues/291
Fix: https://github.com/scenic-views/scenic/issues/380
Fix: https://github.com/scenic-views/scenic/issues/261

calebhearth commented:
@dnamsons @leoplct @joshfester Would you like to give this a shot for the multiple dbs and/or adapters implementation?

@sullerandras @smdern @coding-chimp @marcreniu @tomasc @huarmenta @vladimirtemnikov @nathanpalmer @jdno  @shageman @hlascelles and probably others, this may also get us closer to some of the various Engines and configurable db/views paths needs so I'd love to get input on that as well!

robotfelix commented:
@calebhearth @dnamsons Great work on this. Supporting multiple databases (prior to any built-in Rails support) was the primary reason we originally forked the gem, and it would be great to see this merged and get back on main.

I've done a bit of local testing of this, and everything around migrations (rollbacks, view paths) seems to work great.

https://github.com/scenic-views/scenic/blob/bf8c8aa76fea925c87ecca7bcb059932a35e69cb/README.md?plain=1#L280-L289

One thing I've encountered is that the README updates seem to suggest that adding explicit configuration is only necessary when using multiple adapters, which seems to be true for migrations, but not 100% true overall.

Without this additional configuration then `Scenic.database(:secondary).refresh_materialized_view('my_materialized_view')` executes the SQL against the _primary_ database, which seems like an easy-to-miss footgun. Digging into it, this seems to be because the default adapter uses a connection from `ActiveRecord::Base`. The safest path might be to suggest always adding the extra initializer and to make `Scenic.database(:secondary)` error if none has been configured.

robotfelix commented:
https://github.com/rails/rails/blob/0e53474dd25bb06cc87c07a75ffec49d3490152c/activerecord/lib/active_record/database_configurations.rb#L138-L149

The only other thing I'd note is that there is a subtle distinction between "default" and "primary". Rails seems to mostly use "primary" when referring to the main database in multiple-database setups, but also seems to use "default" within its code.

I'm not confident to say if any of the new code here is using "default" when it should be using "primary", but it's worth noting that in a slightly pathological case like the below, I understand that `:primary` takes precedence over `:default` as the default/automatic `ActiveRecord::Base` database.

```yaml
development:
  default:
    adapter: postgresql
    url: ...
  primary:
    adapter: postgresql
    url: ...
```

calebhearth committed:
Bump CI Ruby/Rails matrix

calebhearth committed:
Raise Rails and Ruby floors in gemspec to match CI

The dependency constraints declared `activerecord`, `activesupport`, and
`railties` at `>= 4.0.0` and Ruby at `>= 2.3.0`, but the CI matrix only
exercises Ruby 3.3+ on Rails 7.2+. The gemspec advertised support
nobody had tested in years.

Bump the floors to match what CI actually verifies. This unblocks
removing version-guard code that exists solely to defend against Rails
versions the gem can no longer install on.

calebhearth committed:
Remove Rails 6.0 multidb version guard

The view generator's `validate_multiple_database_support` Thor task and
`Materializable#validate_rails_version_for_multiple_databases!` raised
on Rails < 6, but the previous commit raised the gem's Rails floor to
7.2. The guard can never trip, so it's dead code that misleads readers
about which Rails versions the gem actually supports.

Also drop the matching "(requires Rails 6.0+)" notes from YARD comments
and the generator option description.

calebhearth committed:
Read views_paths from database config hash

The view generator's --database branch tried to read views_paths via
db_config.respond_to?(:views_paths), but ActiveRecord::DatabaseConfigurations::HashConfig
only exposes built-in keys (migrations_paths and similar) as accessors;
arbitrary settings like views_paths live unaccessed in
configuration_hash. The respond_to? check therefore always returned
false, and the configured views_paths setting was silently ignored —
the generator always fell through to the db/<database>_views convention.

Read the value via configuration_hash[:views_paths] instead, which is
the documented way to access non-built-in configuration entries. Cover
the three HashConfig shapes (configured views_paths, no views_paths,
nil config) at the generator-spec level so the regression is caught
without going through the acceptance pipeline.

Step 3 only fixes the generator's directory-creation path. The
runtime Definition#path is still hard-coded to the convention; that
gets fixed alongside the resolver work in a later commit, at which
point the SQL file itself will land at the configured path.

calebhearth committed:
Make AR configs_for lookup opt-in for the view generator

The generator called ActiveRecord::Base.configurations.configs_for on
every run, defaulting to name: "primary" when --database wasn't set.
That meant every single-DB invocation paid the cost of an AR config
lookup it never needed, and any silent failure mode in that lookup
(e.g. flat-style database.yml without an explicit primary block) hid
legitimate misconfigurations behind a fall-through to convention.

Skip the lookup entirely when database is nil or :default, and only
consult configs_for when --database actually points at a non-default
database. Single-DB users see zero behavior change and zero AR config
calls, while multi-DB users still get configured views_paths and
migrations_paths via the same code path. Drop the now-unused
database_name, conventional_views_path, and conventional_migration_path
helpers and inline the convention paths at the call site.

calebhearth committed:
Convert Definition database arg to a keyword

Definition.new took the database identifier as a third positional
parameter, which made it easy to misread call sites and would have
forced future additions (e.g. an explicit views_paths) into more
positional arguments. Make it a keyword instead so each call site
states which database it's targeting and so future fields are added
without reshuffling call signatures.

Update the four call sites in Statements and the view generator, plus
the spec mocks that expected the old positional shape.

calebhearth committed:
Extract Scenic::DatabasePaths resolver

The view generator and the runtime Definition were drifting toward
two separate copies of "where do view SQL files and migrations live
for this database?" logic. The generator-side check learned about
configuration_hash[:views_paths] in a previous commit, but Definition
was still hard-coded to convention; Codex's adversarial review
flagged this as the architectural punt that would compound entropy
if not centralized.

Introduce Scenic::DatabasePaths as a single source of truth: every
consumer (view generator now, Definition and SchemaDumper soon)
calls into it for views_path, migrations_path, and config_name.
Translation between Scenic's :default symbol and Rails' "primary"
config name lives in one place, so no other code has to know about
the asymmetry.

Migrate the view generator to consume the resolver. Definition and
SchemaDumper remain on their own path-resolution code paths for now;
they get migrated in their own commits where the change in behavior
is testable.

calebhearth committed:
Default Scenic statements to the active AR connection

create_view, drop_view, update_view, and replace_view defaulted to
:default whenever the database keyword wasn't given, so a migration
running under db:migrate:secondary still routed Scenic operations
through the primary adapter unless the author remembered to thread
database: :secondary through every statement. The branch's whole
multi-DB story leaned on that explicit kwarg, which is exactly the
abstraction Codex's adversarial review flagged for removal — every
generated migration, model helper, and command-recorder path was
about to cement the kwarg as load-bearing.

Infer the database from the connection's pool.db_config.name when
the kwarg is omitted; explicit database: still overrides. Database
identity now comes from the connection that's already running, with
:default as a defensive fallback when pool/db_config aren't
available (specs, custom adapters). DatabasePaths gains a reverse
lookup so the symbol/string translation lives in one place.

This shrinks the surface area later steps need to touch: the
generator stops needing to emit redundant database: :secondary in
secondary-DB migrations, command-recorder reverts inherit the
inference, and operations from non-primary connections route
naturally without the user threading the override everywhere.

calebhearth committed:
Plumb views_paths through Definition

Definition built its full path from a hard-coded "db/views" or
"db/<database>_views" convention regardless of any views_paths the
host app had set in database.yml. Statements made things worse by
pre-stripping :default to nil before instantiating Definition,
duplicating a normalization that Definition could already handle.
A migration with views_paths configured for a secondary database
silently looked for SQL files in the wrong directory and raised
Errno::ENOENT.

Have Definition accept an optional views_path kwarg and otherwise
resolve the directory through Scenic::DatabasePaths, so the
generator's directory-creation path and Definition's read path
agree by construction. Drop the redundant :default-to-nil strip
in Statements; the database identifier flows through unchanged
and Definition's resolver handles it natively.

Statements no longer rebuilds the symbol-vs-nil mapping at every
call site, and the next consumer that needs to read views_paths
(SchemaDumper) will reuse the same resolver instead of growing a
third copy.

calebhearth committed:
Resolve SchemaDumper adapter from the dumped connection

dumpable_views_in_database asked Scenic.database (no argument) for
its view list, which always meant the default adapter regardless of
which database the dumper was bound to. Rails runs the dumper once
per database during db:schema:dump and friends, so a multi-DB app
silently produced an empty (or wrong) secondary_schema.rb — exactly
the round-trip failure the multi-database story was supposed to
fix.

Map @connection.pool.db_config.name back to a Scenic database
identifier via the shared resolver, then look up the matching
adapter through Scenic.database. Any database registered with
Scenic.configure now dumps from its own connection, and primary
behaviour stays unchanged because :default falls out of "primary"
naturally.

calebhearth committed:
Route generated model helpers through the model's own connection

Generated materialized-view models emitted Scenic.database.refresh_...
which always resolved to the global default adapter, so a model
generated with --database=secondary would refresh against the primary
database — the wrong relation, or worse, a coincidentally-named one
on the wrong server. Rails 6+'s --database=secondary already
generates the model under SecondaryRecord, which knows its own
connection; routing through Scenic.database threw that information
away.

Add Scenic.adapter_for(connectable) and have the generated model
template call it with self, so refresh and populated? use whichever
connection the class is bound to. ApplicationRecord-based models keep
talking to the primary, SecondaryRecord-based models hit the
secondary, and the generated code stops needing to know about the
adapter registry at all.

calebhearth committed:
Validate --database against configured database names

A typo in --database (e.g. --database=secondry) silently produced
files under db/secondry_views and a migration referencing a
non-existent database, with the failure deferred until somebody
tried to run the migration. Generators should fail fast when given
a name that doesn't match anything in database.yml.

Validate the option against the configured config names at the
moment the generator reads it, raising ArgumentError with the
unknown name and the full list of configured databases. The
validation is memoized so it doesn't re-run for every accessor
call, and :default still passes through without consulting AR's
configurations.

calebhearth committed:
Raise on unknown Scenic database

Scenic.database(name) used to fall back to :default when the name
wasn't registered. A typo in a migration like database: :secondry
silently routed DDL to the primary database with no error, which is
the worst possible failure mode for multi-database operations — the
view ends up on the wrong server with no signal.

Switch the registry to a plain Hash and raise Scenic::UnknownDatabaseError
when an unregistered name is requested. The error message names the
unknown database, lists what is configured, and points at the most
likely cause when it fires during application boot (an initializer
calling Scenic.database before Scenic.configure has run). The
:default adapter is always seeded so single-database setups continue
to work without configuration.

ActiveSupport::OrderedOptions was the only thing standing between a
typo and the silent fallback, and the gem wasn't using its method-call
ergonomics anywhere — a plain Hash is more honest about the data
shape and removes the OrderedOptions tax.

calebhearth committed:
Query view dependencies via the injected connection

Postgres::Views#tsorted_views called ::Scenic.database.execute when
gathering dependent-view metadata, so any Views instance that wasn't
the global default adapter still queried pg_depend on the primary
connection. For a secondary-database adapter that meant computing
dependency graphs against the wrong server — at best empty results,
at worst "relation does not exist" on refresh.

Use the connection that was passed into Views.new instead. The class
already exposes connection as an attr_reader, and views_from_postgres
was already going through it; tsorted_views just needs to do the
same.

calebhearth committed:
Stop emitting redundant database kwarg in generated migrations

Scenic statements now infer the database from the active AR
connection, so a migration that lives under db/secondary_migrate and
runs via db:migrate:secondary already targets the secondary adapter
without any explicit kwarg. Continuing to emit `database: :secondary`
just locked in a redundant API: every generated migration body
referenced a Scenic database identifier that the host app might not
have registered, and after the configuration switch to
raise-on-unknown a plain `rake db:migrate:secondary` failed where it
used to silently fall back.

Drop the database emission from the create/update_view options so
generated migrations stay shape-agnostic. The kwarg is still
available for users who need to override the inferred connection
explicitly; we just don't bake it into every fresh migration.

calebhearth committed:
Read views from the dumper's connection directly

The previous SchemaDumper rewrite resolved a Scenic adapter through
Scenic.configuration.databases keyed by the dumper's connection
name, but multi-DB apps don't always register a Scenic adapter for
every database — they often just rely on database.yml plus the
default Postgres adapter. After the configuration moved to
raise-on-unknown, that lookup raised whenever a non-default
connection was being dumped, breaking db:migrate's automatic
schema-dump pass.

Skip the registry indirection: query Scenic::Adapters::Postgres::Views
directly with the dumper's @connection. The dumper just needs to
know "what views exist on this connection," which the connection
itself answers without any configuration. Customization remains
available for migrations through the registry; the schema dump is a
read-only inspection that doesn't need it.

Acknowledged limitation: this hard-codes the Postgres adapter inside
the dumper. Scenic itself only ships Postgres support today, so this
is a limitation today only insofar as third-party adapters
(scenic-mysql_adapter, scenic-sqlserver_adapter, etc.) would need
their own dumper integration. Worth tracking before any major
release.

calebhearth committed:
Fall back to a connection-built adapter for unregistered databases

Statements inferred the database from the active connection but
still resolved the adapter through Scenic.database, which raised
when the host app hadn't registered an adapter for the inferred
name. Multi-DB apps that just use database.yml without a
Scenic.configure block lost the ability to run scenic statements
under db:migrate:secondary at all.

Distinguish the two reasons we look up an adapter: explicit
database: kwargs are user-controlled and should still raise on
typos, but inferred-from-connection lookups are AR-validated and
should "just work." When the inferred name isn't registered, build
a Postgres adapter on the fly using the migration's own connection
adapter as the connectable. The result hits the same connection AR
already chose for the migration, which is the only sane default.

calebhearth committed:
Clean up multi-database acceptance spec

The acceptance spec mutated the dummy app's working tree without
guaranteed cleanup, used a loose Dir glob that would match leaked
files from previous runs, and carried a commented-out duplicate of
the next migrate command. Tighten the assertions, run cleanup in
after(:each) so a mid-test failure can't contaminate the next run,
and drop the dead comment.

calebhearth committed:
Cover command_recorder reverts that carry a database kwarg

Forward-path multi-DB statements were thoroughly tested but the
reversal path wasn't, even though every revert flows through
StatementArguments and silently relied on the kwarg surviving the
options clone in invert_version / remove_version. Add coverage so a
future change to the inversion logic can't drop the database kwarg
without a failing test.

calebhearth committed:
Replace FakeAdapter with verified instance_double

The hand-rolled FakeAdapter spy duplicated what `instance_double` and
`have_received` already provide, with one important loss: nothing
verified that its method signatures stayed in sync with
Scenic::Adapters::Postgres. Adding a method to the real adapter
would silently leave the fake stale, and the routing tests would
keep passing while the real contract drifted.

Switch the routing specs to instance_double so adapter API drift
trips a NoMethodError rather than going unnoticed. Carry the routing
context into a fresh Scenic::Configuration that is stubbed in via
allow(Scenic).to receive(:configuration), so the tests no longer
mutate global config and don't need an after-each hook to restore
it.

calebhearth committed:
Update README and YARD for the new multi-database contract

The previous prose advertised --database as the abstraction users
work with — every example threaded `database:` through every
statement, and the registration story was a footnote. The new
contract is connection-first, with `database:` reserved as an
override. Rewrite the section to lead with that, drop the
"requires Rails 6.0" note now that the gem floor is 7.2, and call
out the rake-task rename that comes with switching to nested
database.yml so adopters don't get surprised.

Restore SecondaryRecord as the registration example, but show the
abstract-class definition and explain its relationship to
ApplicationRecord so the snippet is copy-pasteable. Document the
configure-before-consume contract that Scenic.database now enforces
via Scenic::UnknownDatabaseError, and update the Scenic.database
YARD comment to match.

calebhearth committed:
Wrap Statements signatures across multiple lines

create_view, update_view, drop_view, and replace_view all crept past
the 100-character line guideline once each gained the new database:
keyword. Wrap each signature one keyword per line so the parameter
list reads cleanly and future additions don't have to choose between
a 130-character line and a separate refactor commit.

calebhearth committed:
Default secondary-database migrations to db/migrate

When --database was set without a configured migrations_paths in
database.yml, the generator put migrations under db/<name>_migrate
by convention. Rails itself defaults multi-database migrations to a
shared db/migrate directory and only honors a separate path when one
is explicitly configured, so Scenic's convention diverged from
Rails' for no good reason — and worse, the divergent path could
collide if a user later added migrations_paths and moved files.

Change the resolver so configured migrations_paths still wins, but
the fallback is db/migrate (matching Rails). The default-database
path keeps its short-circuit so the opt-in semantics from the
generator still hold.

calebhearth committed:
Drop extraneous blank line in schema dumper spec

calebhearth committed:
cleanup
